Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+- Some tests needed to be updated (since a component's nested handlers can't be accessed via `component.content as? SomeHandlerType`)
+- `AnyComponent` is replaced by `AnyEndpointNode` and `AnyEndpointProvidingNode`
+- Neither `content` nor `handle()` have a default implementation anymore.
+  You have to provide an implementation for your components to compile.
+  (The exception here is if you manually add a `typealias {Content|Response} = Never`, in which case we do still provide a default implementation. The idea is that, if you want a component to be unusable, and crash if ever accessed, you have to explicitly specify this.)
+
+
+
+### Benefits from switching to a more fine-grained protocol structure:
+
+**Type Safety**
+
+With the current implementation, we'd need to rely heavily on runtime checks to validate the DSL structure and rejct invalid web services.
+
+By using protocols to define some of these component concepts, we can make use of the type system to enforce these requirements.
+
+This directly benefits both the user (ie the person implementing a web service), and us:
+
+- The user will, in some cases, get compile-time errors when trying to compile incorrectly implemnented components, or ill-formed DSL constructs
+- We can, when dealing with DSL nodes, always know for a fact their specific type and functionalities and won't have to do any dynamic or runtime-metadata based checks to determine that information
+
+
+
+
+### Issues with the current implementation in [experimental/dsl-node-protocols][branch_url]
+
+**Function builder complexity**  
+The issue here is that DSL constructs can now consist of two types instead of just one.  
+This means that, whereas previously the only variable (ie, difference between `buildBlock` functions) was the number of parameters, we now also have to take the potential types into account.
+
+
+Example: in the current function builder, the `buildBlock()` signatures are:
+
+```swift
+(Component) -> some Component
+(Component, Component) -> some Component
+(Component, Component, ..., Component) -> some Component
+```
+
+Since the new builder has to support two types, and any permutation between them, the function builder becomes:
+
+```swift
+(EndpointNode) -> some EndpointProvidingNode
+(EndpointProvidingNode) -> some EndpointProvidingNode
+
+(EndpointNode, EndpointNode) -> some EndpointProvidingNode
+(EndpointNode, EndpointProvidingNode) -> some EndpointProvidingNode
+(EndpointProvidingNode, EndpointNode) -> some EndpointProvidingNode
+(EndpointProvidingNode, EndpointProvidingNode) -> some EndpointProvidingNode
+
+...
+```
+
+Basically, we have to support all potential permutations/mixings of endpoint- and non-endpoint nodes in the function builder.
+This also means that the number of generated functions isn't `O(n)` anymore, but `O(n^2)`.  
+(it's been a while since i took GAD, so i could be wrong here...)
+
+
+
+
+**Certain use cases necessitate code duplication**
```

Comment:
I have the feeling that this would also be solved by making `Handler`/`EndpointNode` conform to `Component`/`EndpointProvidingNode`. I suspect it would still be possible to write modifiers that only work on `Handler`s/`EndpointNode`s which might even be disable as e.g. an `operation(...)` (https://github.com/Apodini/Apodini/blob/develop/Documentation/Modifiers/Operation.md) modifier wouldn't make too much sense on a `Group`.

Reply:
I tried this (having `Handler` conform to `Component`) and the issue is that, since the `Modifier` protocol itself needs to inherit from `Component`, you lose type information after applying one modifier to a component.

Example from the GuardTests:
```swift
Text("Hallo")            // type: Handler
    .operation(.read)    // type: Component
    .guard(TestGuard())  // type: Component
    .operation(.create)  // type: Component
```

The first modifier returns an `OperationModifier<Text>` which, via conformance to `Modifier`, conforms to `Component` but not to `Handler`, meaning we're "downgrading" the `Handler` type to a mere `Component`.

Ideally, we'd be able to fix this by writing something along the lines of
```swift
extension Modifier: Handler where Self.ModifiedComponent: Handler {
    public typealias Response = ModifiedComponent.Response
}
```
but that doesn't compile since you can't declare protocol inheritance in an extension.